### ISSUE TITLE
Automatic media player embedding doesn't work for PeerTube, SoundCloud and other places

### ISSUE THREAD

**Issue by bkil** [friendica/friendica#10019]
- [x] I have searched open and closed issues for duplicates

### Bug Description

If I share or embed a PeerTube video link, the resulting post only contains an empty frame in the middle along with the video link and description. The same happens for SoundCloud.

Embedding Vimeo or YouTube by URL doesn't work either, but they do work inside Vimeo and YouTube video tags **if given by ID**, respectively.

### Steps to Reproduce

1. Create a new post
2.  Click the "link or media" button at the bottom and paste the URL (or type in some [embedding BBCode](https://github.com/friendica/friendica/blob/46f9e9351e57a00ffa7f939704f37a61c425acab/doc/BBCode.md#embed) manually like `[video]` or `[audio]`)
3.  Press Share

Actual Result:

No video or audio player will appear, only an empty frame in the middle.

In case of YouTube or Invidious URL's, only a thumbnail will appear in place of a video player. While this is documented, it is unexpected behavior.

Expected Result:

I would expect an audio or video player to appear.

As a workaround, I could embed a deeplink to the mp4 file on the PeerTube instance and fill in the rich preview content by hand, but that's not acceptable for most users.

### Platform Info

Friendica Version: 2021.03-rc 

Friendica Source: https://nerdica.net/friendica 

PHP version:

SQL version:

**Comment by bkil** [comment]
Maybe not the root cause, but I'm missing `oembed_video.tpl` mentioned here:
- https://github.com/friendica/friendica/blob/46f9e9351e57a00ffa7f939704f37a61c425acab/src/Content/OEmbed.php#L196

Also, changing video URL to video ID for Vimeo and YouTube might be done here, though ideally it should also work for the "link or media" button on the form:
- https://github.com/friendica/friendica/blob/46f9e9351e57a00ffa7f939704f37a61c425acab/src/Content/Text/BBCode/Video.php#L42

**Comment by annando** [comment]
`oembed_video.tpl` should be in `view/templates/`.

**Comment by bkil** [comment]
I've proceeded with debugging. The regression is probably caused by #9675. I haven't bisected it, but from memory, it worked up to around that time.

After adding the URL of my PeerTube instance to `URI.SafeIframeRegexp` in `BBCode.php`, a new post did contain a working video player. Surely a better solution will be needed.

**Comment by mikauskas** [comment]
@bkil @annando Looks like it is the same problem first reported on 17 of December 2020 #9666. As far as I know, @MrPetovan introduced by disabling IFrame [https://github.com/friendica/friendica/issues/9666#issuecomment-775453982](url)
Similar problems reported on #9912 #9922 #9962 #9915 #9666 #9935
Nothing is being done in an effort to fix all those bugs, except silencing me on many of them.
I think trying to use OEmbed for all so many different types of embedding is condemned. Especially for Peertube, because each Peertube node has a different domain name. Moreover, because Friendica is not really designed for hosting and streaming videos, I think there must be made better integration of Friendica with Peertube. I think there must be another way, especially, because Peertube itself is part of Fediverse, a kind of local network. I suggest contacting Peetube developers and asking for advice, would contact myself, but don't know what kind of questions to ask [https://github.com/Chocobozzz/PeerTube/issues](url)

**Comment by bkil** [comment]
After some more testing, the situation seems a bit more complicated. It actually embeds a different kind of YouTube player depending on whether you use the "link or media" button or ID via BBCode, but that's for a different day.

However, look at the question from a bird's eye view, it does sound reasonable to limit the scope of embedding external resources from a security & tracking perspective. But from the same perspective, we should actually **not embed YouTube, Twitter and Facebook videos** et al directly with an iframe, because in that case Alphabet will track every user who has the post showing up in their timeline. Many users of Friendica came to the platform in order to avoid such tracking, and reintroducing it is kind of a step backwards.

A privacy-aware workaround would be to cache & proxy the thumbnail image of such media and only load the embedded player (and autoplay) after the user clicks on or hovers over the image. If we chose to implement such a workaround, this could probably be generalized to work for any kind of iframes, including PeerTube and SoundCloud ones. A drawback would be that embedded players that do not support autoplay would require the user to click twice in such cases, though I think this could be gated through a user privacy setting as well.

Although, from a privacy perspective, if an instance admin operates a whole portfolio of services, including Friendica, PixelFed, PeerTube, NextCloud, Funkwhale and so on, it would be perfectly reasonable to enable any and all iframes from the subdomains operated by the admin. Such prefix, regexp or domain whitelists could even be added as admin panel settings. However, this would only enable showing these iframes to local users, so basically making the fediverse less useful, hence this workaround does not scale.

The best possible workaround that I could think of that has less latency and/or enable autoplay would involve basically shipping one embeddable web player for each popular decentralized media application on the Friendica instance and only using the data files from the linked remote servers. In this case, again only trusted web content would be embedded. This would be a challenge from the side of packaging and having a bit service fingerprinting logic, but would otherwise be just a matter of a few static asset files. Additionally, copyright questions of these codes should also be thought out appropriately.

This would surely not be allowed for players of proprietary platforms, but I wouldn't really worry about those.

I've checked out your linked issues, but I can't give a good recommendation for generalized rich preview OEmbeds. The issue here is that even if the HTML to embed would be cached within the database, it could reference further external resources that could also lead to tracking by outside parties. Hence, HTML normalization, filtering and further proxying of images and CSS inlining would be needed there, but this all sounds like a major undertaking. Not sure if anyone else has already implemented a library for that, though.

**Comment by mikauskas** [comment]
The wast majority of those who consider coming to the platform don't give a dam about the possibility of such kind of tracking. Those who care enough can start their own server. All they care is illegal censorship. A 100 million of them moved to Telegram this January alone. Although Telegram is not a social network at all, there are channels and it has no problems with embedding anything. On other hand latest versions of Freindica work like a messenger from MsDOS era - text only. Why not remove an attachment and video buttons already? Would be a simpler user interface. Even Mastodon with 500 characters limit is much better than Friendica since 2021.01. Why not shut down all nodes? Would be no tracking, or security vulnerabilities guaranteed!
Peertube is open-source, so no problems with copyrights, others like Youtube somewhat working or possible to fix with OEmbed. 
If we worry too much about all Fediverse nothing will be done.
If you have ideas on how to fix all that mess I think you will be very welcome, because @MrPetrovan himself said
> None of the current core developers know how to handle video properly' 

[https://github.com/friendica/friendica/issues/9910#issuecomment-774713091](url)

**Comment by annando** [comment]
@bkil the idea behind Peertube is that the videos are provided by several torrent links. Currently we ignore this and we only display the MP4 video that directs to the original server. In the future we should change this behaviour of course.

Also for this reason I would recommend sharing the whole original video post. This would distribute also the whole bunch of torrent links and possibly other systems are doing it better. Sharing the original post (via the native reshare, not the commented reshare) does preserve all the information.

Concerning the sharing of Youtube videos: It does work on my machine, which is on the latest RC. We had some problem with this, but this had been fixed. Could you provide some sample links for different audio and video content, so we can test it with the same data?

**Comment by bkil** [comment]
Here are some sample posts (see the URL under the post):
- SoundCloud URL in with the media button (empty frame) https://nerdica.net/display/a85d7459-1760-4944-62bb-2d6753458418
- SoundCloud URL in audio BBCode (empty frame) https://nerdica.net/display/a85d7459-1160-4945-1c51-f69789485939
- YouTube URL with media button (static thumbnail): https://nerdica.net/display/a85d7459-3960-477c-8072-0ba756923323
- YouTube URL with URL button/url BBCode (nothing): https://nerdica.net/display/a85d7459-1260-4946-7c63-425519758802
- YouTube URL in video (or youtube) BBCode (embeds something, but playback error):  https://nerdica.net/display/a85d7459-1960-4946-fc0b-406677717135
- Youtube ID in youtube BBCode (this one works): https://nerdica.net/display/a85d7459-1060-4947-b4a0-765557879382

As a humanoid user, one would expect that the same media player would get inserting with any of the above methods, especially the ones supported by widgets. This is incidentally how it works on some other big platforms.

As I've mentioned above, YouTube and Vimeo are hardcoded in `SafeIframeRegexp`, so I doubt that other providers would work.

**Comment by annando** [comment]
@bkil I did some improvements, so that the embedding via the "link or media" button should work now with media files.

**Comment by annando** [comment]
@bkil BTW: Please use the `[audio]` and `[video]` elements only when directly linking a media link. These links won't work with - for example - web pages that contain some media.

**Comment by bkil** [comment]
On the contrary, adding PeerTube links worked splendidly inside `[video]` tags before December. It still works correctly to this day on my own instance if I add my PeerTube instance to `SafeIframeRegexp`.

The documentation also mentions this as a supported feature:
* https://github.com/friendica/friendica/blob/46f9e9351e57a00ffa7f939704f37a61c425acab/doc/BBCode.md#embed

> [video]url[/video] | Where *url* can be an url to youtube, vimeo, soundcloud, or other sites wich supports oembed or opengraph specifications.
> [video]Video file url[/video] [audio]Audio file url[/audio] | Full URL to an ogg/ogv/oga/ogm/webm/mp4/mp3 file. An HTML5 player will be used to show it.

Could we please reopen this issue? I've listed lots of things that you did not address.
* If you don't expect to fix PeerTube embedding, at least provide a workaround to insert the video file URL instead automatically.
* The above enumerated YouTube player/usablitiy errors would also need to be fixed or at least tracked.
* If SoundCloud isn't supported anymore, documentation (and expectation) needs to be changed, although this sounds a bit backwards.

**Comment by AlfredSK** [comment]
For me Vimeo and Youtube works (player is loading, playback possible) when I completely **disable** oembed in admin > site > policy > no oembed... (w/o any exceptions).
Soundcloud is only showing a preview of the page (with image and link). No player - with or w/o oembed.
Peertube videos are showing an empty frame (player not loading) and a link to the peertube content below that frame.

**Comment by annando** [comment]
Yeah, I just had a look at the BBCode parser. We do call the same functionality for the audio, video, vimeo and youtube elements that checks for OEmbed. Only for some specific extensions the `[video]` will directly lead to just the `<video>` element.

**Comment by mikauskas** [comment]
Yes, seriously, @bkil is right, why this issue has been closed? I can't see any improvements for Peertube. In my understanding, Peetube is the main issue here? @MrPetrovan works like Tony Fauci - when this is such a good product, that you want it to give to people, as quickly as possible, why you should wait until testers dot all i-s and cross all t-s?
`[video]https://gorf.tube/videos/watch/fc5b2ebc-771f-451f-82a8-92c9bb2b0e52[/video]`

Expected result, works great on v2020.09.1
[https://fv91.friendica.online/display/bc07ef10-7860-4cd4-7aa6-32c873151982](url)
![Screenshot_2021-03-13 Friendica Social Network Network](https://user-images.githubusercontent.com/76782677/111034735-ed440a80-840e-11eb-9f36-bb26402eb373.png)

Real result v2021.03-rc - 1412 
`https://quality.friendica.lt/display/70b6ca56-7160-4cd7-a296-27a794569172`

![Screenshot_2021-03-13 Friendica Social Network testeris Friendica Social Network](https://user-images.githubusercontent.com/76782677/111034851-5cb9fa00-840f-11eb-91de-03d72334645e.png)

**Comment by MrPetovan** [comment]
@bkil The issue was closed automatically because @annando mentioned in #10038 that it `Fixes #10019` and the PR was merged, reopening.

**Comment by annando** [comment]
@bkil my fix handled an issue with OEmbed content that had falsely declared itself as audio or video (which I found during testing your issue)

But you had been right with https://github.com/friendica/friendica/issues/10019#issuecomment-794584733 concerning the origin of the problem with the empty iframe.

This is something I would like to delegate to @MrPetovan.

**Comment by MrPetovan** [comment]
I have two ideas about that:
- Disabled the iframe source filter in HTMLPurifier when OEmbed is enabled for any domain.
- Add the Oembed allowlisted domains to the filter.

**Comment by annando** [comment]
These were my two thoughts as well. Only thing is: I haven't found the right configuration for this. Also I think it would be good to not have an empty iframe when it had been blocked. Question is if this is configurable as well.

**Comment by MrPetovan** [comment]
HTML Purifier is supposed to remove iframes if they don't have an `src` attribute, but at the moment with "No Oembed rich content" enabled and an empty allowlist, iframes still display with their content, and without "No OEmbed rich content" no iframe appears, which is the exact opposite from the expected behavior.

It doesn't appear related to caching since I tried with different URLs, I have to dive into OEmbed once more.

**Comment by annando** [comment]
@MrPetovan have a look at my PR #10044 - this seems to work on my system.

**Comment by MrPetovan** [comment]
I kept looking into it and it's more complicated than expected, or at least the settings have to be reworded.

The OEmbed domain allowlist is only used for `type=rich` content. However, it isn't the only OEmbed type that uses `<iframe>` HTML tags. `type=video` like PeerTube video also provide embed URL to be used in an iframe. Taking the strict meaning of the allowlist setting, it isn't a good way to enable iframes for non-rich content like PeerTube videos.

And then there's the issue where HTML Purifier doesn't remove source-filtered iframes since we provide a width and a height. I'll keep at it.

**Comment by MrPetovan** [comment]
Here are my final words on the subject before going to bed, I'll submit a PR tomorrow: 

1. The "No OEmbed rich content" setting is irrelevant to this issue, we should filter out untrusted iframe sources whether it's enabled or not. It isn't only a matter of privacy, it is first and foremost a matter of security, and OEmbed produces iframes with types other than `rich`.
2. In this context, the "Allowed OEmbed domains" setting feature can and probably should be extended to trust iframe source domains beyond the current behavior linked with the state of the "No OEmbed rich content".

**Comment by annando** [comment]
Whatever solution we will finally have, it should achieve a solution where people can allow **any** content to be embedded (when they want to, they should be able doing so) or some selected content. Concerning the selected content a user won't be able to know the exact path to the embedded player (like `https://w.soundcloud.com/player/`)

Have a look at my latest commit to the PR. I changed the behaviour concerning OEmbed data that contains a HTML object.

Edit: the settings `no_oembed_rich_content` and `allowed_oembed` are used in `isAllowedURL()` to determine for which addresses OEmbed data should be fetched. From my point of view this is primarily meant for embedding rich content. Also I wouldn't want to change the general behaviour at this time of the development. (I  made some changes in my PR where I had been unsure if they are okay for this stage, but I hope I thought of every edge case here)

**Comment by MrPetovan** [comment]
I saw what you did with `unset($oembed->html);` and I've tried that as well, but we're touching about how specific the "No OEmbed rich content" setting is. It's currently only relevant to the `type=rich` OEmbed responses. For example PeerTube will respond with an `<iframe>` and `type=video`.

I'm not necessarily against this change, but this means we have to change the description of both settings instead of just "Allowed OEmbed domains". I like that it would make embeds nicer by default if the "No OEmbed rich content" setting is enabled and that we don't need both the OEmbed and the `getSiteinfo` requests for this case.

>  it should achieve a solution where people can allow any content to be embedded (when they want to, they should be able doing so)

The "they" in this sentence is ambiguous, so I will try to explain why we have to filter out untrusted iframe sources. If "they" is "admins", they probably don't want users to embed harmful websites in iframes in their posts. Unfortunately there is no reliable way of doing so than with a trusted list.

If "they" is users, it includes potential malicious actors and "when they want to, they should be able to [embed **any** content]" takes another dimension which also justifies the iframe source filtering.

**Comment by annando** [comment]
Several points here: I would define this `video` type as "rich" as well since it provides some HTML. And this provided HTML which will be displayed in an inline frame is exactly what this settings controls.

This combination of OEmbed and OpenGraph (and others) is something that we can focus on in the next version, I think. I don't really like this recursive call of the functions. But In this late state of development I wouldn't wanted to change too significant settings.

Concerning my "they" sentence: I don't really like to patronize the users/admins without any possibility to change this behaviour. Especially on single users instances the administrator is the only user and should be able to allow any inline frame from OEmbed sources. We should clearly emphasize the security issues. But we should allow it.

And also I think that the settings `no_oembed_rich_content` and `allowed_oembed` clearly control if "rich" data (which of course includes embedded videos as well) should be allowed in general or for some selected systems. And since "rich" data always is displayed in inline frames I don't see a reason for separate configurations. I also see this inline frame filtering as some kind of safety belt/airbag. Means: an inline frame should only be created for selected sources and the filtering should only prevent content from malicious sources.

**Comment by MrPetovan** [comment]
> Especially on single users instances the administrator is the only user and should be able to allow any inline frame from OEmbed sources.

Unfortunately at the moment we do not have a way to distinguish the source of the iframes. If we allow inline frames from any source, it will include non-OEmbed sources possibly from remote posts. And yes, security is almost always inconvenient. I do not mind an "allow all" mode but it should be an additional configuration setting disabled by default.

> And also I think that the settings no_oembed_rich_content and allowed_oembed clearly control if "rich" data (which of course includes embedded videos as well) should be allowed in general or for some selected systems. And since "rich" data always is displayed in inline frames I don't see a reason for separate configurations.

I agree with you but this isn't what these settings currently do in the code nor exactly how they are described. We do not need separate configurations but we would need to make their description match their behavior.

**Comment by MrPetovan** [comment]
So it seems there are three main tasks to be done:
1. Offer a way to allow domains in OEmbed iframes by slightly expanding the "Allow OEmbed domains" setting scope, this would be enough to satisfy this issue for this release. (#10046)
2. Rework the "No OEmbed rich content" setting into a "No OEmbed" setting that would skip OEmbed entirely and rely on "getInfosite" instead. This should be done after this release.
3. Add a "Allow all domains" setting that would completely bypass the ifram source filter. This should also be done after this release.

**Comment by mikauskas** [comment]
> Unfortunately at the moment we do not have a way to distinguish the source of the iframes

Perhaps administrators could whitelist specific Peertube domains?

**Comment by mikauskas** [comment]
Just noticed that when Peertube video [https://gorf.tube/videos/watch/fc5b2ebc-771f-451f-82a8-92c9bb2b0e52](url) being posted on Discourse, their editor displays an error
`Sorry, we were unable to generate a preview for this web page, because the following oEmbed / OpenGraph tag could not be found: description `

Perhaps it is something that must be fixed on Peetube software?

Discourse still manages to generate fairly decent preview using code

```
<header class="source">
      <img src="https://gorf.tube/client/assets/images/favicon.png?626d50cbacafc17fc902ad3037a7f00295db49d8" class="site-icon" width="16" height="16">
      <a href="https://gorf.tube/videos/watch/fc5b2ebc-771f-451f-82a8-92c9bb2b0e52" target="_blank" rel="noopener">GORF Tube</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:133/75;"><img src="https://gorf.tube/lazy-static/previews/fc5b2ebc-771f-451f-82a8-92c9bb2b0e52.jpg" class="thumbnail" width="133" height="75"></div>
<h3><a href="https://gorf.tube/videos/watch/fc5b2ebc-771f-451f-82a8-92c9bb2b0e52" target="_blank" rel="noopener">CELEB EXPOSES FAUCI’S MISINFORMATION - The Highwire</a></h3>
  </article>
  <div class="onebox-metadata">
       </div>
  <div style="clear: both"></div>
  ```
  

![Screenshot_2021-03-15 Postų redagavimo testavimas - Puslapio atsiliepimai - Infomanija(2)](https://user-images.githubusercontent.com/76782677/111146044-e50bdc80-8580-11eb-9c2e-1131b8175b72.png)

**Comment by annando** [comment]
Concerning https://github.com/friendica/friendica/issues/10019#issuecomment-798931999 I would like to rework this whole part in the next release. Means:

- A single class to fetch page related data, using sub classes for OEmbed, OpenGraph and JsonLd. This class would return the page data as class variables like OEmbed already is doing.
- Settings that allow "rich" content from selected or all pages. (We could use the existing config variables but should consider renaming the description and revert the default value)

**Comment by MrPetovan** [comment]
> > Unfortunately at the moment we do not have a way to distinguish the source of the iframes
> 
> Perhaps administrators could whitelist specific Peertube domains?

Yes, this is what has been done in #10046 through the now-called "Trusted third-party domains" where it used to be "Allowed OEmbed domains".

**Comment by mikauskas** [comment]
> > Perhaps administrators could whitelist specific Peertube domains?
> 
> Yes, this is what has been done in #10046 through the now-called "Trusted third-party domains" where it used to be "Allowed OEmbed domains".

Yes! Working with majority links. Just needed to resave old posts. @MrPetovan you are a good man. Thanks a lot. 
It also solves the majority of problems with #9962

**Comment by MrPetovan** [comment]
@bkil Are you satisfied with https://github.com/friendica/friendica/pull/10046 ?

**Comment by tobiasd** [comment]
What is the status of this issue?

**Comment by bkil** [comment]
The proposed workarounds didn't sound right to me, but let me upgrade my instance and try it how it is working now.

**Comment by xundeenergie** [comment]
a peertube-video got by activitypub plays only sound, but no video is seen in the player.
I have 2021.06rc i think, friendica grabs the wrong url for the videosource.


Friendica has for example this https://kino.schuerz.at/static/webseed/cbb6b465-8ed6-4a00-aa32-2dd37d4e4413-0.mp4 video-source for the video... the webseed is not correct.

This is the video for the webseed above

https://kino.schuerz.at/videos/watch/cbb6b465-8ed6-4a00-aa32-2dd37d4e4413

The working embedding video link is

https://kino.schuerz.at/videos/embed/cbb6b465-8ed6-4a00-aa32-2dd37d4e4413

**Comment by xundeenergie** [comment]
I found, i had an audio-only-version created automatically in peertube. The url for the embedding video from peertube grabs obviously the first "video", which ist the audio-only... not the "embed" video

**Comment by tio-trom** [comment]
Ok. So from my tests. If I want to share a Peertube video on my instance with or without "No OEmbed rich content " : 

1. Share the link directly https://videos.trom.tf/videos/watch/955d6be8-d8e7-430b-9079-61043f5f1cae it grabs the metadata about the link as such:

![image](https://user-images.githubusercontent.com/38043315/127510026-b3fca68e-f923-4d97-b4a7-d067447b7822.png)


2. Click the media button and add the url above:
![image](https://user-images.githubusercontent.com/38043315/127510149-b6b08537-dc00-481c-9226-aa92f6c4959b.png)


3. Click the media button and add the .mp4 video from Peertube:
![image](https://user-images.githubusercontent.com/38043315/127510318-eaa85aec-b0cf-4b4c-8dda-058e679fc85a.png)


4. Add the embed code:
![image](https://user-images.githubusercontent.com/38043315/127510446-e5dfe78a-39a7-40a6-bf63-3d4adc2de7bd.png)


I imagine someone who shares a Peertube video link simply grabs the main link and shares it. This will share it as any other website. I would have expected, since it is activiypub, to be able to grab the video itself and embed into the post. This is only possible if you grab the download url of a Peertube video, which is not ideal in my view.

Another thing, the OEmbed .... enabled or disabled seems to make no difference on my instance. I run the latest stable Friendica.

**Comment by MrPetovan** [comment]
@tio-trom The last PR improved the display of shared PeerTube posts specifically from your server 😄

**Comment by tio-trom** [comment]
Oh wow so it will automatically pickup the .mp4 and embed it when a peertube video is shared? That's really fantastic honestly! Awesome job!

**Comment by MrPetovan** [comment]
Yes, and thank you! It's on develop only at the moment.

**Comment by MrPetovan** [comment]
@bkil Can you please pull the RC branch and verify this issue has been resolved?

**Comment by bkil** [comment]
Okay, will check tomorrow.

**Comment by tobiasd** [comment]
Any news? Or should we move this ticket to the December release?

**Comment by bkil** [comment]
So the status is that I upgraded my test instance to 21.07, but couldn't get it up to 21.09rc. After hours of debugging and bisecting, I couldn't get it to migrate between between database version 1424->1425, stopping due to an ill query that would upgrade apcontact.

```
[16-Sep-2021 16:18:15 UTC] PHP Fatal error: Uncaught PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'uri-id' in 'group statement' in src/Database/Database.php:530
```
I tried to fix up the SQL queries and do the steps manually, then it produced foreign key errors when proceeding to `fcontact`, so I didn't proceed. Has anybody seen this before?

**Comment by annando** [comment]
When calling `bin/console dbstructure update -f` doesn't fix your database, then please open a separate issue with your upgrade problem, because this is unrelated to this issue here.

**Comment by tobiasd** [comment]
So, using the URL of a video on Peertube in the video or attachment tag, now results like this for me on 2021.09-rc

![grafik](https://user-images.githubusercontent.com/433490/134137043-ef6917ff-a6c7-4fe4-989f-43ef18b6dbf4.png)

but opens the player on the Peertube side. Not sure if this is intended. Could also be the effect of OEmbed not activated on my node.

**Comment by annando** [comment]
I don't think that we have got an issue here. We can create posts that contain links to a Peertube video, like for example this one: https://pirati.ca/display/ec054ce7-6062-9206-b741-6c2680514654

And you can also simply reshare a peertube video like you would reshare a text post. So from my side I don't see an issue here.

**Comment by bkil** [comment]
Thank you for the link. I can only see a preview image of the video, not a video player itself being embedded.

**Comment by annando** [comment]
Yes. But I don't think that we should embed it. When we want to share a Peertube video then we should share that video, like we share a post with a test or picture.

**Comment by bkil** [comment]
In contrast, YouTube **is** embedded whether you like it or not. Your timeline & network feed sends a bunch of tracker pings to a bunch of big proprietary players. The only way around this at present is to fully block it on the networking level.

In contrast, why can't a Friendica operator (or the Friendica user even) enumerate in the settings the list of trusted PeerTube/PixelFed instances that they trust to serve embedded content?

Also, if we aim lower than P2P, you could just as well embed the direct link to the mobile mp4 file hosted on PeerTube with a regular `<video>` widget - that way you wouldn't need to whitelist any JavaScript call at all.

And also, we opt to only show this proxied and cached cover image until the user clicks on it and only then create the `<video>` widget. This won't ping any third party server at all with the video stream until the user opts to play the video, so it could work without a whitelist. A banner/warning/caption might be shown under the video widget about the origin, though.

**Comment by MrPetovan** [comment]
> In contrast, YouTube **is** embedded whether you like it or not. Your timeline & network feed sends a bunch of tracker pings to a bunch of big proprietary players. The only way around this at present is to fully block it on the networking level.

I had to add the various Youtube domains as "trusted third-party domains" in the Admin settings in order for embeds to work, I'm not sure it's "whether you like it or not".

**Comment by annando** [comment]
Whatever we will do with this, I don't consider it a bug. It's more a feature request.

**Comment by foss-** [comment]
Certainly a feature request and milestone should be removed.

**Comment by bkil** [comment]
You mean via `URI.SafeIframeRagexp`? Before a user joins an instance, how can we tell what is the value of such a setting? Is SoundCloud working now and the other corner cases enumerated in the first few comments? I can't test YouTube.

**Comment by bkil** [comment]
Okay, so I retested Vimeo. I have the same issue as reported in the top post:

* You will only get the clickable title preview if you add the URL in either `[vimeo]` or `[video]` BBCode.
* You will get a rich preview (a cover image, title and prefix of the description) if you use the `link or media` button, but no video player.
* You will get an embedded video player widget if you add the numerical suffix of the URL in `[vimeo]` BBCode.
* You will get **both** the embedded video player widget and the rich preview _side by side_ if you just mention the vimeo URL within the post body. This is surely the best in terms of usability but it begs the question: why not do exactly the same in all four cases?

I couldn't get a SoundCloud player widget to appear in any combination despite the fact that it does offer an oembed iframe in both JSON and XML.

**Comment by tio-trom** [comment]
Hey so I was trying to write a post on friendica and use inline videos but I am kinda lost how to. For youtube it is as simple as adding the url inline. Then magic happens. The sad thing is for peertube or invidious this does not and I have no clue how to add these in a post.

**Comment by foss-** [comment]
Just ran into this sharing an invidious link. If player embedding is out of scope due to having to add all instances as trusted third-party domains as per https://github.com/friendica/friendica/issues/10019#issuecomment-1140264494 should at least a preview image be pulled for the link preview when adding an invidious URL to the post?

Unsure which code style is to be used or expected by friendica and what currently should be showing in the post.

Talked to MrPetovan about this but this seems not trivial to resolve for invidious.

Wondering if the invidious discussion should be split into a separate issue since supporting better link recognicion of an alternative frontend is yet another feature request than improving PeerTube video embedding.

**Comment by tio-trom** [comment]
For Peertube the embed code looks like this:

`<iframe title="ZDay: How to Grow a Saner Society. A Lifetime Adventure --- by Tio" width="560" height="315" src="https://videos.trom.tf/videos/embed/61cf7395-26a4-4820-b17e-7edd6bec67df" frameborder="0" allowfullscreen="" sandbox="allow-same-origin allow-scripts allow-popups"></iframe>`

And responsive as:

`<div style="position: relative; padding-top: 56.25%;"><iframe title="ZDay: How to Grow a Saner Society. A Lifetime Adventure --- by Tio" width="100%" height="100%" src="https://videos.trom.tf/videos/embed/61cf7395-26a4-4820-b17e-7edd6bec67df" frameborder="0" allowfullscreen="" sandbox="allow-same-origin allow-scripts allow-popups" style="position: absolute; inset: 0px;"></iframe></div>`

I think it would also make sense to have a video-embed button instead of the link/media button that seems to just add a link with the URL that you insert.

**Comment by MrPetovan** [comment]
For PeerTube the simpler way to share a video is natively resharing the post itself. They can be imported in Friendica b pasting the video page link in the search box.

SoundCloud could be handled specifically, like YouTube and Vimeo.

**Comment by tio-trom** [comment]
I see but I was thinking more about the ability to embed inline, so between text. Like we can do for Youtube ;)

**Comment by MrPetovan** [comment]
We match the domain name for YouTube to be able to use their embed code. We can't do that for PeerTube without fully probing the remote URL, which would be very costly in the BBCode replacement routine. Better probing the video page URL in the search context where it's meant to be done.

**Comment by tio-trom** [comment]
I see. So basically because youtube always has 1 single domain youtube.com it is easy to embed via the BBCOde, but since Peertube can have any URL there is no way to know if the URL is a Peertube or not?

If so would it be easier to have a `[peertube]URL[/peertube]` in BBCode that basically adds the URL to this:

`<iframe width="560" height="315" src="URL" frameborder="0" allowfullscreen="" sandbox="allow-same-origin allow-scripts allow-popups"></iframe>`

So it is up to the people to add a proper Peertube URL same as it is when using other BBCode snippets. 

EDIT: Oh well but yeah `src=` can be different from the actual URL.... tricky

**Comment by MrPetovan** [comment]
If we can extract the video identifier from the main video page URL, we should able to implement a specific `[peertube]` BBCode but I’d still rather have these tags for third-party services with no ActivityPub support.

**Comment by haheute** [comment]
Sometimes I find myself thinking I like the embedded player on mastodon.. (for peertube it works, too) but I trust your decision (as I have no idea how much work that would be and what's the more practical or logical way to deal with posted video links).
